File: services/agent-worker/bm_scout_worker/runtime_env.py
from __future__ import annotations

import os
import re
from pathlib import Path

DEFAULT_ENV_FILES = (".env.local", ".env")
ENV_KEY_PATTERN = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*$")


def load_local_env_files(cwd: str | Path | None = None, files: tuple[str, ...] = DEFAULT_ENV_FILES) -> dict[str, list[str]]:
    root = Path(cwd or Path.cwd())
    loaded_files: list[str] = []
    loaded_keys: list[str] = []

    for file_name in files:
        path = root / file_name
        if not path.exists():
            continue
        loaded_files.append(file_name)
        for line in path.read_text(encoding="utf-8").splitlines():
            parsed = parse_env_line(line)
            if not parsed:
                continue
            key, value = parsed
            if key in os.environ:
                continue
            os.environ[key] = value
            loaded_keys.append(key)

    return {"files": loaded_files, "keys": loaded_keys}


def parse_env_line(line: str) -> tuple[str, str] | None:
    trimmed = line.strip()
    if not trimmed or trimmed.startswith("#") or "=" not in trimmed:
        return None
    key, raw_value = trimmed.split("=", 1)
    key = key.strip()
    if not ENV_KEY_PATTERN.match(key):
        return None
    return key, unquote_env_value(raw_value.strip())


def unquote_env_value(value: str) -> str:
    if len(value) >= 2 and value.startswith('"') and value.endswith('"'):
        return value[1:-1].replace("\\n", "\n").replace('\\"', '"').replace("\\\\", "\\")
    if len(value) >= 2 and value.startswith("'") and value.endswith("'"):
        return value[1:-1]
    comment_match = re.search(r"\s#", value)
    if comment_match:
        return value[: comment_match.start()].strip()
    return value.strip()

File: src/server/runtime-env.ts
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";

export interface LoadedEnvFiles {
  files: string[];
  keys: string[];
}

const DEFAULT_ENV_FILES = [".env.local", ".env"];

export function loadLocalEnvFiles(cwd = process.cwd(), files = DEFAULT_ENV_FILES): LoadedEnvFiles {
  const loadedFiles: string[] = [];
  const loadedKeys: string[] = [];

  for (const file of files) {
    const path = join(cwd, file);
    if (!existsSync(path)) continue;
    loadedFiles.push(file);
    const content = readFileSync(path, "utf8");
    for (const line of content.split(/\r?\n/)) {
      const parsed = parseEnvLine(line);
      if (!parsed) continue;
      const [key, value] = parsed;
      if (process.env[key] !== undefined) continue;
      process.env[key] = value;
      loadedKeys.push(key);
    }
  }

  return { files: loadedFiles, keys: loadedKeys };
}

export function parseEnvLine(line: string): [string, string] | null {
  const trimmed = line.trim();
  if (!trimmed || trimmed.startsWith("#")) return null;
  const index = trimmed.indexOf("=");
  if (index <= 0) return null;
  const key = trimmed.slice(0, index).trim();
  if (!/^[A-Za-z_][A-Za-z0-9_]*$/.test(key)) return null;
  const rawValue = trimmed.slice(index + 1).trim();
  return [key, unquoteEnvValue(rawValue)];
}

function unquoteEnvValue(value: string): string {
  if (value.length >= 2 && value.startsWith("\"") && value.endsWith("\"")) {
    return value.slice(1, -1).replace(/\\n/g, "\n").replace(/\\"/g, "\"").replace(/\\\\/g, "\\");
  }
  if (value.length >= 2 && value.startsWith("'") && value.endsWith("'")) {
    return value.slice(1, -1);
  }
  const commentIndex = value.search(/\s#/);
  return (commentIndex >= 0 ? value.slice(0, commentIndex) : value).trim();
}
